packages/lesswrong/components/posts/TableOfContents/FixedPositionToC.tsx

```diff
@@ -257,25 +259,97 @@ const FixedPositionToc = ({tocSections, title, heading, onClickSection, displayO
+      const postContentRect = postContentElement.getBoundingClientRect();
+      
+      // Calculate the offset of post content from the top of the container
+      const postContentOffsetInContainer = postContentRect.top - containerRect.top + container.scrollTop;
+      const postContentHeight = postContentElement.offsetHeight;
+      
+      // Calculate where the viewport BOTTOM is relative to the post content
+      // This matches the "fixed ToC" calculation in usePostReadProgress
+      const viewportBottomInContainer = container.scrollTop + container.clientHeight;
+      const distanceFromPostTopToViewportBottom = viewportBottomInContainer - postContentOffsetInContainer;
+      
+      // The scroll percent represents how far the viewport bottom has traveled through the post
+      const scrollPercent = Math.max(0, Math.min(100, (distanceFromPostTopToViewportBottom / postContentHeight) * 100));
+      
+      // Calculate the window height as a percentage of the post content
+      const viewportHeightAsPercentOfPost = (container.clientHeight / postContentHeight) * 100;
+      
+      if (readingProgressBarRef.current) {
+        // The progress bar uses flex, so scrollPercent represents the portion above the indicator
+        // But we need to adjust it to represent where the viewport TOP is
+        const adjustedScrollPercent = Math.max(0, scrollPercent - viewportHeightAsPercentOfPost);
+        readingProgressBarRef.current.style.setProperty("--scrollAmount", `${adjustedScrollPercent}%`);
+        
+        // Set window height as percentage of the progress bar container
+        const progressBarHeight = readingProgressBarRef.current.offsetHeight;
+        const windowHeightInPixels = (viewportHeightAsPercentOfPost / 100) * progressBarHeight;
+        readingProgressBarRef.current.style.setProperty("--windowHeight", `${windowHeightInPixels}px`);
+      }
+    };
+    update();
+    container.addEventListener('scroll', update);
+    return () => container.removeEventListener('scroll', update);
+  }, [scrollContainerRef, disableProgressBar, readingProgressBarRef]);
+
   const jumpToAnchor = (anchor: string) => {
     if (isServer) return;
 
+    const anchorElement = document.getElementById(anchor);
+    if (!anchorElement) return;
+
+    const container = scrollContainerRef?.current;
+
+    // Prefer scrolling inside the provided scroll container if available
+    if (container) {
+      const containerRect = container.getBoundingClientRect();
+      const anchorRect = anchorElement.getBoundingClientRect();
+      const offsetInsideContainer = anchorRect.top - containerRect.top;
+
+      const targetScrollTop = container.scrollTop + offsetInsideContainer - (container.clientHeight * 0.2);
+
+      container.scrollTo({
+        top: targetScrollTop, // keep heading ~20% from top
+        behavior: 'smooth',
+      });
+
+      // Update URL hash for consistency
+      delete query.commentId;
```

Comment:
This isn't safe, since `query` might be used elsewhere.  You should create a shallow clone without `commentId` instead